File: TinkerRocketApp/TinkerRocketApp/Models/UnitFormatter.swift
//
//  UnitFormatter.swift
//  TinkerRocketApp
//
//  Centralized display-unit formatting (#160).  All telemetry, .bin/.csv
//  logs, and the wire protocol stay SI (canonical) — these helpers convert
//  at render time per the user's chosen UnitSystem, so logged data is never
//  altered and analysis scripts are unaffected.
//
//  Methods default to UnitSystem.current so non-View code (FlightAnnouncer,
//  TelemetryData computed props) can format too.  Views pass their
//  @AppStorage("unitSystem") value so the UI re-renders when the toggle flips.
//

import Foundation

enum UnitSystem: String, CaseIterable, Identifiable {
    case metric
    case imperial

    var id: String { rawValue }

    var label: String {
        switch self {
        case .metric:   return "Metric (SI)"
        case .imperial: return "Imperial"
        }
    }

    /// Persisted choice, readable from non-View code.  Shares the
    /// "unitSystem" UserDefaults key with the @AppStorage in SettingsView.
    static var current: UnitSystem {
        UnitSystem(rawValue: UserDefaults.standard.string(forKey: "unitSystem") ?? "") ?? .metric
    }
}

// Exact conversion factors.
private let feetPerMeter = 1.0 / 0.3048
private let feetPerMile = 5280.0
private let standardGravity = 9.80665   // m/s² per g

enum UnitFormatter {

    // MARK: - Altitude
    //
    // Never switches to miles: rocket apogees routinely exceed 5280 ft and
    // "1.2 mi" is useless for altitude.  Horizontal distances use distance().

    static func altitude(_ meters: Double, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:
            return meters >= 1000
                ? String(format: "%.2f km", meters / 1000.0)
                : String(format: "%.0f m", meters)
        case .imperial:
            return String(format: "%.0f ft", meters * feetPerMeter)
        }
    }

    // MARK: - Horizontal distance (promotes to km / mi when large)

    static func distance(_ meters: Double, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:
            return meters >= 1000
                ? String(format: "%.1f km", meters / 1000.0)
                : String(format: "%.0f m", meters)
        case .imperial:
            let feet = meters * feetPerMeter
            return feet >= feetPerMile
                ? String(format: "%.2f mi", feet / feetPerMile)
                : String(format: "%.0f ft", feet)
        }
    }

    // MARK: - Speed

    static func speed(_ mps: Double, decimals: Int = 1, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:   return String(format: "%.\(decimals)f m/s", mps)
        case .imperial: return String(format: "%.\(decimals)f ft/s", mps * feetPerMeter)
        }
    }

    static func speedUnit(_ system: UnitSystem = .current) -> String {
        system == .metric ? "m/s" : "ft/s"
    }

    // MARK: - Acceleration (g in imperial)

    static func acceleration(_ mps2: Double, decimals: Int = 2, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:   return String(format: "%.\(decimals)f m/s\u{00B2}", mps2)
        case .imperial: return String(format: "%.\(decimals)f g", mps2 / standardGravity)
        }
    }

    /// Unit suffix only — for table headers / IMU row labels that format the
    /// triplet of axis values separately.
    static func accelerationUnit(_ system: UnitSystem = .current) -> String {
        system == .metric ? "m/s\u{00B2}" : "g"
    }

    /// Scale an acceleration triplet into the display unit (m/s² or g) without
    /// formatting — callers lay out the three axes themselves.
    static func accelerationValue(_ mps2: Double, system: UnitSystem = .current) -> Double {
        system == .metric ? mps2 : mps2 / standardGravity
    }

    // MARK: - Temperature

    static func temperature(_ celsius: Double, decimals: Int = 1, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:   return String(format: "%.\(decimals)f \u{00B0}C", celsius)
        case .imperial: return String(format: "%.\(decimals)f \u{00B0}F", celsius * 9.0 / 5.0 + 32.0)
        }
    }

    // MARK: - Spoken (FlightAnnouncer voice callouts)

    static func spokenAltitude(_ meters: Double, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:   return "\(Int(meters.rounded())) meters"
        case .imperial: return "\(Int((meters * feetPerMeter).rounded())) feet"
        }
    }

    static func spokenSpeed(_ mps: Double, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:   return "\(Int(mps.rounded())) meters per second"
        case .imperial: return "\(Int((mps * feetPerMeter).rounded())) feet per second"
        }
    }

    static func spokenDistance(_ meters: Double, system: UnitSystem = .current) -> String {
        switch system {
        case .metric:   return "\(Int(meters.rounded())) meters"
        case .imperial: return "\(Int((meters * feetPerMeter).rounded())) feet"
        }
    }

    // MARK: - Raw numeric conversions (charts, input fields)

    static func metersToDisplay(_ meters: Double, system: UnitSystem = .current) -> Double {
        system == .metric ? meters : meters * feetPerMeter
    }

    static func displayToMeters(_ value: Double, system: UnitSystem = .current) -> Double {
        system == .metric ? value : value / feetPerMeter
    }

    static func mpsToDisplay(_ mps: Double, system: UnitSystem = .current) -> Double {
        system == .metric ? mps : mps * feetPerMeter
    }

    static func displayToMps(_ value: Double, system: UnitSystem = .current) -> Double {
        system == .metric ? value : value / feetPerMeter
    }

    // MARK: - CSV column conversion (FlightChartView)
    //
    // Logs are canonical SI; the chart converts at render time keyed on the
    // trailing "(unit)" token in the column header.  Non-convertible columns
    // (deg, deg/s, uT, Pa, mA, V, %, ms, flags) pass through unchanged.

    /// Returns the display label + converted values for a plotted CSV column.
    static func convertSeries(column: String, values: [Double],
                              system: UnitSystem = .current) -> (label: String, values: [Double]) {
        guard system == .imperial else { return (column, values) }

        switch columnUnit(column) {
        case "m":
            return (relabel(column, to: "ft"), values.map { $0 * feetPerMeter })
        case "m/s":
            return (relabel(column, to: "ft/s"), values.map { $0 * feetPerMeter })
        case "m/s2":
            return (relabel(column, to: "g"), values.map { $0 / standardGravity })
        case "C":
            return (relabel(column, to: "\u{00B0}F"), values.map { $0 * 9.0 / 5.0 + 32.0 })
        default:
            return (column, values)
        }
    }

    /// Extract the unit token inside the trailing "(...)" of a column header.
    private static func columnUnit(_ column: String) -> String? {
        guard let open = column.lastIndex(of: "("),
              let close = column.lastIndex(of: ")"),
              open < close else { return nil }
        return String(column[column.index(after: open)..<close])
    }

    /// Replace the trailing "(unit)" of a column header with a new unit.
    private static func relabel(_ column: String, to unit: String) -> String {
        guard let open = column.lastIndex(of: "(") else { return column }
        return String(column[..<open]) + "(\(unit))"
    }
}

File: TinkerRocketAndroid/core/session/src/main/kotlin/com/tinkerbug/tinkerrocket/session/UnitFormatter.kt
package com.tinkerbug.tinkerrocket.session

import java.util.Locale
import com.tinkerbug.tinkerrocket.protocol.formatFixed

/**
 * Display-unit formatting, ported from the iOS `UnitFormatter` (#160) —
 * format-string for format-string, so the two apps print identical values in
 * either system.  Wire, log and CSV stay SI everywhere; conversion happens
 * only at the point of display.  The spoken twin is [SpokenUnits].
 */
public object UnitFormatter {

    private const val FEET_PER_METER: Double = 1.0 / 0.3048
    private const val FEET_PER_MILE: Double = 5280.0
    private const val STANDARD_GRAVITY: Double = 9.80665

    // #1086: round ties-to-even (printf / iOS), not HALF_UP. String.format
    // rounds `%f` HALF_UP, so a frame carrying an exact .5 printed one metre
    // higher on Android than iOS. `fmt` here is always "%.<N>f<suffix>" (every
    // call site below), so split the decimals and the unit suffix off and run
    // the number through the shared formatFixed (BigDecimal HALF_EVEN, the same
    // routine the CSV uses). Falls back to String.format for any fmt that is
    // not that shape, which none currently is.
    private val fixedFmt = Regex("""%\.(\d+)f(.*)""", RegexOption.DOT_MATCHES_ALL)
    private fun f(fmt: String, v: Double): String {
        val m = fixedFmt.matchEntire(fmt)
            ?: return String.format(Locale.ROOT, fmt, v)
        val decimals = m.groupValues[1].toInt()
        val suffix = m.groupValues[2]
        return formatFixed(v, decimals) + suffix
    }

    // ── Altitude ─────────────────────────────────────────────────────────
    // Never switches to miles: rocket apogees routinely exceed 5280 ft and
    // "1.2 mi" is useless for altitude.  Horizontal distances use distance().

    public fun altitude(meters: Double, system: UnitSystem): String = when (system) {
        UnitSystem.METRIC ->
            if (meters >= 1000) f("%.2f km", meters / 1000.0) else f("%.0f m", meters)
        UnitSystem.IMPERIAL -> f("%.0f ft", meters * FEET_PER_METER)
    }

    /** Base altitude unit label ("m" / "ft") — no km/mi promotion. */
    public fun altitudeUnit(system: UnitSystem): String =
        if (system == UnitSystem.METRIC) "m" else "ft"

    /** Scale an altitude into the display unit without formatting (iOS
     *  `metersToDisplay`) — for callers with their own format string. */
    public fun altitudeValue(meters: Double, system: UnitSystem): Double =
        if (system == UnitSystem.METRIC) meters else meters * FEET_PER_METER

    /** Inverse of [altitudeValue] (iOS `displayToMeters`) — parse a user
     *  entry in the display unit back to canonical meters. */
    public fun altitudeToMeters(display: Double, system: UnitSystem): Double =
        if (system == UnitSystem.METRIC) display else display / FEET_PER_METER

    // ── Horizontal distance (promotes to km / mi when large) ─────────────

    public fun distance(meters: Double, system: UnitSystem): String = when (system) {
        UnitSystem.METRIC ->
            if (meters >= 1000) f("%.1f km", meters / 1000.0) else f("%.0f m", meters)
        UnitSystem.IMPERIAL -> {
            val feet = meters * FEET_PER_METER
            if (feet >= FEET_PER_MILE) f("%.2f mi", feet / FEET_PER_MILE) else f("%.0f ft", feet)
        }
    }

    // ── Speed ────────────────────────────────────────────────────────────

    public fun speed(mps: Double, system: UnitSystem, decimals: Int = 1): String = when (system) {
        UnitSystem.METRIC -> f("%.${decimals}f m/s", mps)
        UnitSystem.IMPERIAL -> f("%.${decimals}f ft/s", mps * FEET_PER_METER)
    }

    public fun speedUnit(system: UnitSystem): String =
        if (system == UnitSystem.METRIC) "m/s" else "ft/s"

    // ── Acceleration (g in imperial) ─────────────────────────────────────

    public fun acceleration(mps2: Double, system: UnitSystem, decimals: Int = 2): String =
        when (system) {
            UnitSystem.METRIC -> f("%.${decimals}f m/s²", mps2)
            UnitSystem.IMPERIAL -> f("%.${decimals}f g", mps2 / STANDARD_GRAVITY)
        }

    public fun accelerationUnit(system: UnitSystem): String =
        if (system == UnitSystem.METRIC) "m/s²" else "g"

    /** Scale an acceleration triplet into the display unit without formatting. */
    public fun accelerationValue(mps2: Double, system: UnitSystem): Double =
        if (system == UnitSystem.METRIC) mps2 else mps2 / STANDARD_GRAVITY

    // ── Temperature ──────────────────────────────────────────────────────

    public fun temperature(celsius: Double, system: UnitSystem, decimals: Int = 1): String =
        when (system) {
            UnitSystem.METRIC -> f("%.${decimals}f °C", celsius)
            UnitSystem.IMPERIAL -> f("%.${decimals}f °F", celsius * 9.0 / 5.0 + 32.0)
        }
}
